### ISSUE TITLE
Seg-fault when clickhouse-backup server is busy

### ISSUE THREAD

**Issue by therealnb** [Altinity/clickhouse-backup#149]
I wrote some code to see how well we handle multiple instances of the backup tool on different ports. Mostly it went well, but there were a few seg faults like this

```
    panic: runtime error: invalid memory address or nil pointer dereference
    [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xebab6c]
    
    goroutine 2345 [running]:
    github.com/AlexAkulov/clickhouse-backup/pkg/backup.moveShadow.func1(0xc0010d63f0, 0x63, 0x0, 0x0, 0x1501760, 0xc0010b6ae0, 0x0, 0x0)
	    /home/travis/gopath/src/github.com/AlexAkulov/clickhouse-backup/pkg/backup/utils.go:39 +0x16c
    path/filepath.walk(0xc0010c8b90, 0x4d, 0x15228c0, 0xc0010e2a90, 0xc0004dd278, 0x0, 0x0)
	    /home/travis/.gimme/versions/go1.14.linux.amd64/src/path/filepath/path.go:380 +0x20c
    path/filepath.walk(0xc0002bac80, 0x34, 0x15228c0, 0xc00108f1e0, 0xc0004dd278, 0x0, 0x0)
	    /home/travis/.gimme/versions/go1.14.linux.amd64/src/path/filepath/path.go:384 +0x2ff
    path/filepath.walk(0xc0002eb7d0, 0x24, 0x15228c0, 0xc00108f040, 0xc0004dd278, 0x0, 0x0)
	    /home/travis/.gimme/versions/go1.14.linux.amd64/src/path/filepath/path.go:384 +0x2ff
    path/filepath.walk(0xc00061f0e0, 0x1e, 0x15228c0, 0xc00108ef70, 0xc0004dd278, 0x0, 0x0)
	    /home/travis/.gimme/versions/go1.14.linux.amd64/src/path/filepath/path.go:384 +0x2ff
    path/filepath.walk(0xc00061f080, 0x19, 0x15228c0, 0xc00108eea0, 0xc0004dd278, 0x0, 0x0)
	    /home/travis/.gimme/versions/go1.14.linux.amd64/src/path/filepath/path.go:384 +0x2ff
    path/filepath.walk(0xc00104d020, 0x17, 0x15228c0, 0xc001055790, 0xc0004dd278, 0x0, 0xc00104cfe0)
	    /home/travis/.gimme/versions/go1.14.linux.amd64/src/path/filepath/path.go:384 +0x2ff
    path/filepath.Walk(0xc00104d020, 0x17, 0xc0004dd278, 0xc00104cfe0, 0x18)
	    /home/travis/.gimme/versions/go1.14.linux.amd64/src/path/filepath/path.go:406 +0xff
    github.com/AlexAkulov/clickhouse-backup/pkg/backup.moveShadow(0xc00104d020, 0x17, 0xc000fa70e0, 0x4c, 0x17, 0x0)
	    /home/travis/gopath/src/github.com/AlexAkulov/clickhouse-backup/pkg/backup/utils.go:32 +0x83
    github.com/AlexAkulov/clickhouse-backup/pkg/backup.CreateBackup(0xc0002903e8, 0x2, 0x0, 0x0, 0x0, 0xc000290430, 0x8, 0xc000290448, 0x8, 0xc000290490, ...)
	    /home/travis/gopath/src/github.com/AlexAkulov/clickhouse-backup/pkg/backup/backup.go:521 +0x7b5
    github.com/AlexAkulov/clickhouse-backup/pkg/server.(*APIServer).httpCreateHandler.func1(0xc000520000, 0xc00006c1fa, 0x2d, 0xc00006c1da, 0x1a)
	    /home/travis/gopath/src/github.com/AlexAkulov/clickhouse-backup/pkg/server/server.go:478 +0x103
    created by github.com/AlexAkulov/clickhouse-backup/pkg/server.(*APIServer).httpCreateHandler
	    /home/travis/gopath/src/github.com/AlexAkulov/clickhouse-backup/pkg/server/server.go:476 +0x3d4
```

This doesn't look like it is particularly related to running multiple clickhouse-backup servers, but it may be so I reported it. The script was like this

```
  import json, threading, time
  from urllib import request, parse
  from urllib.error import HTTPError
  
  # Start multiple servers with 
  # for p in 7172 7173 7174 7175 7176 ; do sudo -u clickhouse S3_BUCKET=multitest API_LISTEN=0.0.0.0:$p /usr/bin/clickhouse-backup server --config=/etc/clickhouse-backup/config.yml & done
  
  ch_server = 'clickhouse.local'
  ch_backup_ports = [7172, 7173, 7174, 7175, 7176]
  tables = ['table1', 'table2', 'table3', 'table4']
  db_name = 'ns_v2'
  last_activity = time.time()
  
  def do_post(url):
    global last_activity
    try:
      last_activity = time.time()
      print(f'POSTING {url}')
      req =  request.Request(url, method='POST')
      resp = request.urlopen(req)
      print(f'POST {url} = {resp.code}, {resp.msg}')
    except HTTPError as http_err:
      e = json.load(http_err.fp)
      print(f'POST {url} = HTTP error occurred: {e["error"]}. Retry in 3s...')
      if 'another operation is currently running' in e["error"]:
        time.sleep(3)
        do_post(url)
    except Exception as err:
      print(f'POST {url} = Unexpected error occurred: {err}')
    last_activity = time.time()
  
  def create_backup(srv, port, table, name):
    def do_create_backup(args):
      url = f'http://{srv}:{port}/backup/create?table={db_name}.{table}&name={name}'
      do_post(url)
    return do_create_backup
  
  def upload_backup(srv, port, name):
    def do_upload_backup(args):
      url = f'http://{srv}:{port}/backup/upload/{name}'
      do_post(url)
    return do_upload_backup
  
  def delete_backup(srv, port, table, name):
    def do_delete_backup(args):
      url = f'http://{srv}:{port}/backup/delete/local/{name}'
      do_post(url)
      url = f'http://{srv}:{port}/backup/delete/remote/{name}.tar.gz'
      do_post(url)
    return do_delete_backup
  
  def clean_shadow(srv, port):
    def do_clean_shadow(args):
      url = f'http://{srv}:{port}/backup/clean'
      do_post(url)
    return do_clean_shadow
  
  def multi_clean():
    for p in ch_backup_ports:
      try:
        t = threading.Thread(target=clean_shadow(ch_server, p), args=(p, ) )
        t.start()
      except Exception as e:
        print(f'Exception: {e}')
  
  def multi_create():
    for p, t in [(p, t) for p in ch_backup_ports for t in tables]:
      try:
        t = threading.Thread(target=create_backup(ch_server, p, t, f'0000{p}{t}'), args=(p, ) )
        t.start()
      except Exception as e:
        print(f'Exception: {e}')
  
  def multi_delete():
    for p, t in [(p, t) for p in ch_backup_ports for t in tables]:
      try:
        t = threading.Thread(target=delete_backup(ch_server, p, t, f'0000{p}{t}'), args=(p, ) )
        t.start()
      except Exception as e:
        print(f'Exception: {e}')
  
  def multi_upload():
    for p, t in [(p, t) for p in ch_backup_ports for t in tables]:
      try:
        t = threading.Thread(target=upload_backup(ch_server, p, f'0000{p}{t}'), args=(p, ) )
        t.start()
      except Exception as e:
        print(f'Exception: {e}')
  
  def multi_delete():
    for p, t in [(p, t) for p in ch_backup_ports for t in tables]:
      try:
        t = threading.Thread(target=delete_backup(ch_server, p, t, f'0000{p}{t}'), args=(p, ) )
        t.start()
      except Exception as e:
        print(f'Exception: {e}')
  
  def wait_until_quiet(qt):
    global last_activity
    print('Wait for a bit...')
    now = time.time()
    while (now - last_activity) < qt:
      print('busy...')
      time.sleep(qt)
      now = time.time()
    print('Quiet.')
    last_activity = time.time()
```

**Comment by AlexAkulov** [comment]
Why do you want to have multiple clickhouse-backups on the same host?

**Comment by therealnb** [comment]
It is just one possible workaround to increase s3 up/download speeds for multiple large backups. It will allow them to happen in parallel.

So far, this is just an investigation. It may be that in the future the rest  may be more parallel and that would be better.

If you can think of reasons we shouldn't use multiple servers on one host, we'd love to hear about them.

Many thanks

**Comment by therealnb** [comment]
I should add that this ticket is more for information, rather than something we want fixed immediately. Retries and systemd will compensate for this.

**Comment by AlexAkulov** [comment]
This panic caused by 'create' command running in parallel. clickhouse-backup manipulates data in the 'shadow' folder and parallel processes were not provided, but I have a way to get around this and the next version should work correctly in this case.

**Comment by therealnb** [comment]
Good to know. Thanks

**Comment by AlexAkulov** [comment]
Fixed in [v1.0.0-alpha1](https://github.com/AlexAkulov/clickhouse-backup/releases/tag/v1.0.0-alpha1)
